### PR TITLE
[MI-3159] Implement OAuth for jira cloud instances

### PR DESCRIPTION
**Pull request by raghavaggarwal2308**
Recreated from an existing PR: https://github.com/mattermost/mattermost-plugin-jira/pull/949

### CODE REVIEW

File: server/client.go [hunk 1]
```diff
@@ -310,7 +310,7 @@ func (client JiraClient) GetSelf() (*jira.User, error) {
 // MakeCreateIssueURL makes a URL that would take a browser to a pre-filled form
 // to file a new issue in Jira.
 func MakeCreateIssueURL(instance Instance, project *jira.Project, issue *jira.Issue) string {
-	u, err := url.Parse(fmt.Sprintf("%v/secure/CreateIssueDetails!init.jspa", instance.GetURL()))
+	u, err := url.Parse(fmt.Sprintf("%v/secure/CreateIssueDetails!init.jspa", instance.GetJiraBaseURL()))
```
Comment (Kshitij-Katiyar) [inline server/client.go:313]:
```suggestion
	url, err := url.Parse(fmt.Sprintf("%v/secure/CreateIssueDetails!init.jspa", instance.GetJiraBaseURL()))
```

File: server/command.go [hunk 2]
```diff
@@ -799,6 +805,44 @@ func executeInstanceInstallCloud(p *Plugin, c *plugin.Context, header *model.Com
 	})
 }
 
+func executeInstanceInstallCloudOAuth(p *Plugin, c *plugin.Context, header *model.CommandArgs, args ...string) *model.CommandResponse {
+	authorized, err := authorizedSysAdmin(p, header.UserId)
+	if err != nil {
+		return p.responsef(header, "%v", err)
```
Comment (Kshitij-Katiyar) [inline server/command.go:811]: Can we use this here
```suggestion
		return p.responsef(header, err.Error())
```

File: server/command.go [hunk 3]
```diff
@@ -799,6 +805,44 @@ func executeInstanceInstallCloud(p *Plugin, c *plugin.Context, header *model.Com
...
+
+	state := flow.State{
+		keyEdition:          string(CloudOAuthInstanceType),
+		keyJiraURL:          jiraURL,
+		keyInstance:         instance,
+		keyOAuthCompleteURL: p.GetPluginURL() + instancePath(routeOAuth2Complete, types.ID(jiraURL)),
+		keyConnectURL:       p.GetPluginURL() + instancePath(routeUserConnect, types.ID(jiraURL)),
+	}
+
+	err = p.oauth2Flow.ForUser(header.UserId).Start(state)
```
Comment (Kshitij-Katiyar) [inline server/command.go:833]: Check this everywhere
```suggestion
	if err = p.oauth2Flow.ForUser(header.UserId).Start(state); err !=nil {
```

File: server/command.go [hunk 4]
```diff
@@ -799,6 +805,44 @@ func executeInstanceInstallCloud(p *Plugin, c *plugin.Context, header *model.Com
...
+	if err != nil {
+		return p.responsef(header, err.Error())
+	}
+
+	channel, _ := p.client.Channel.GetDirect(header.UserId, p.conf.botUserID)
```
Comment (Kshitij-Katiyar) [inline server/command.go:838]: if the second variable is err please check for error then

File: server/instance_cloud_oauth.go [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+var _ Instance = (*cloudOAuthInstance)(nil)
+
+const (
+	JiraScopes        = "read:jira-user,read:jira-work,write:jira-work"
+	JiraScopesOffline = JiraScopes + ",offline_access"
+	JiraResponseType  = "code"
+	JiraConsent       = "consent"
+)
+
+func (p *Plugin) installCloudOAuthInstance(rawURL string, clientID string, clientSecret string) (string, *cloudOAuthInstance, error) {
```
Comment (Kshitij-Katiyar) [inline server/instance_cloud_oauth.go:51]: Check this everywhere
```suggestion
func (p *Plugin) installCloudOAuthInstance(rawURL , clientID , clientSecret string) (string, *cloudOAuthInstance, error) {
```

File: server/instance_cloud_oauth.go [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+	jiraURL, err := utils.CheckJiraURL(p.GetSiteURL(), rawURL, false)
+	if err != nil {
+		return "", nil, err
+	}
+	if !utils.IsJiraCloudURL(jiraURL) {
+		return "", nil, errors.Errorf("`%s` is a Jira server URL, not a Jira Cloud", jiraURL)
```
Comment (Kshitij-Katiyar) [inline server/instance_cloud_oauth.go:57]:
```suggestion
		return "", nil, errors.Errorf("`%s` is a Jira server URL instead of Jira Cloud", jiraURL)
```

File: server/instance_cloud_oauth.go [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+
+	instance := &cloudOAuthInstance{
+		InstanceCommon:   newInstanceCommon(p, CloudOAuthInstanceType, types.ID(jiraURL)),
+		MattermostKey:    p.GetPluginKey(),
+		JiraClientID:     clientID,
+		JiraClientSecret: clientSecret,
+		JiraBaseURL:      rawURL,
+	}
+
+	err = p.InstallInstance(instance)
```
Comment (Kshitij-Katiyar) [inline server/instance_cloud_oauth.go:68]: same

File: server/instance_cloud_oauth.go [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+	updatedToken, err := tokenSource.Token()
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "error getting token from token source")
+	}
+
+	if updatedToken.RefreshToken != currentToken.RefreshToken {
+		connection.OAuth2Token = updatedToken
+
+		// Store this new access token & refresh token to get a new access token in future when it has expired
+		err = ci.Plugin.userStore.StoreConnection(ci.Common().InstanceID, connection.MattermostUserID, connection)
```
Comment (Kshitij-Katiyar) [inline server/instance_cloud_oauth.go:101]: same

File: server/instance_cloud_oauth.go [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+func (ci *cloudOAuthInstance) GetURL() string {
+	return "https://api.atlassian.com/ex/jira/" + ci.JiraResourceID
+}
+
+func (ci *cloudOAuthInstance) GetJiraBaseURL() string {
+	return ci.JiraBaseURL
+}
+
+func (ci *cloudOAuthInstance) GetManageAppsURL() string {
+	// TODO
```
Comment (Kshitij-Katiyar) [inline server/instance_cloud_oauth.go:162]: Find out what is TODO here and explain if possible ?

File: server/instance_cloud_oauth.go [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+	return fmt.Sprintf("%s/plugins/servlet/webhooks", ci.GetURL())
+}
+
+func (ci *cloudOAuthInstance) GetMattermostKey() string {
+	return ci.MattermostKey
+}
+
+func (ci *cloudOAuthInstance) getJiraCloudResourceID(client http.Client) (string, error) {
+	request, err := http.NewRequest(
```
Comment (Kshitij-Katiyar) [inline server/instance_cloud_oauth.go:176]: We dont usually call any API directly see if they have functions like call or callJson and use them if possible ?

File: server/instance_cloud_oauth.go [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+		"GET",
+		"https://api.atlassian.com/oauth/token/accessible-resources",
+		nil,
+	)
+	if err != nil {
+		return "", fmt.Errorf("failed getting request")
```
Comment (Kshitij-Katiyar) [inline server/instance_cloud_oauth.go:182]:
```suggestion
		return "", fmt.Errorf("failed to get the request")
```

File: server/instance_cloud_oauth.go [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+	}
+
+	response, err := client.Do(request)
+	if err != nil {
+		return "", fmt.Errorf("failed getting accessible resources: %s", err.Error())
```
Comment (Kshitij-Katiyar) [inline server/instance_cloud_oauth.go:187]: same

File: server/instance_cloud_oauth.go [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+	}
+
+	defer response.Body.Close()
+	contents, err := io.ReadAll(response.Body)
+	if err != nil {
+		return "", fmt.Errorf("failed read accesible resources response: %s", err.Error())
```
Comment (Kshitij-Katiyar) [inline server/instance_cloud_oauth.go:193]:
```suggestion
		return "", fmt.Errorf("failed to read accessible resources response: %s", err.Error())
```

File: server/instance_cloud_oauth.go [hunk 14]
```diff
@@ -0,0 +1,209 @@
...
+	}
+
+	var resources JiraAccessibleResources
+	err = json.Unmarshal(contents, &resources)
+
```
Comment (Kshitij-Katiyar) [inline server/instance_cloud_oauth.go:198]:
```suggestion
```

File: server/instance_cloud_oauth.go [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+	}
+
+	defer response.Body.Close()
+	contents, err := io.ReadAll(response.Body)
+	if err != nil {
+		return "", fmt.Errorf("failed read accesible resources response: %s", err.Error())
+	}
+
+	var resources JiraAccessibleResources
+	err = json.Unmarshal(contents, &resources)
```
Comment (Kshitij-Katiyar) [inline server/instance_cloud_oauth.go:197]: same

File: server/instance_cloud_oauth.go [hunk 16]
```diff
@@ -0,0 +1,209 @@
...
+
+	if err != nil {
+		return "", errors.Wrap(err, "failed to unmarshal JiraAccessibleResources")
+	}
+
+	// We return the first resource ID only
```
Comment (Kshitij-Katiyar) [inline server/instance_cloud_oauth.go:203]:
```suggestion
	// We just return the first resource ID only
```

File: server/kv.go [hunk 17]
```diff
@@ -498,6 +499,15 @@ func (store *store) LoadInstanceFullKey(fullkey string) (Instance, error) {
 		ci.Plugin = store.plugin
 		return &ci, nil
 
+	case CloudOAuthInstanceType:
+		ci := cloudOAuthInstance{}
+		err = json.Unmarshal(data, &ci)
```
Comment (Kshitij-Katiyar) [inline server/kv.go:504]: same

File: server/setup_flow.go [hunk 18]
```diff
@@ -554,15 +584,51 @@ func (p *Plugin) submitCreateCloudInstance(f *flow.Flow, submission map[string]i
...
+	jiraURL, _ := submission["url"].(string)
+	if jiraURL == "" {
+		return "", nil, nil, errors.New("no Jira cloud URL in the request")
+	}
+	jiraURL = strings.TrimSpace(jiraURL)
+	if jiraOrgRegexp.MatchString(jiraURL) {
+		jiraURL = fmt.Sprintf("https://%s.atlassian.net", jiraURL)
+	}
+
+	clientID, _ := submission["client_id"].(string)
```
Comment (Kshitij-Katiyar) [inline server/setup_flow.go:611]: Constants for these 3 params

File: server/setup_flow.go [hunk 19]
```diff
@@ -554,15 +584,51 @@ func (p *Plugin) submitCreateCloudInstance(f *flow.Flow, submission map[string]i
...
+		keyEdition:          string(CloudOAuthInstanceType),
+		keyJiraURL:          jiraURL,
+		keyInstance:         instance,
+		keyOAuthCompleteURL: p.GetPluginURL() + instancePath(routeOAuth2Complete, types.ID(jiraURL)),
+		keyConnectURL:       p.GetPluginURL() + instancePath(routeUserConnect, types.ID(jiraURL)),
+	}, nil, nil
+}
+
+func (p *Plugin) submitCreateCloudOAuthInstance(f *flow.Flow, submission map[string]interface{}) (flow.Name, flow.State, map[string]string, error) {
+	jiraURL, _ := submission["url"].(string)
```
Comment (Kshitij-Katiyar) [inline server/setup_flow.go:602]: check for this error

File: server/setup_flow.go [hunk 20]
```diff
@@ -554,15 +584,51 @@ func (p *Plugin) submitCreateCloudInstance(f *flow.Flow, submission map[string]i
...
+	if jiraURL == "" {
+		return "", nil, nil, errors.New("no Jira cloud URL in the request")
```
Comment (Kshitij-Katiyar) [inline server/setup_flow.go:604]:
```suggestion
		return "", nil, nil, errors.New("no Jira cloud URL present in the request")
```

File: server/setup_flow.go [hunk 21]
```diff
@@ -554,15 +584,51 @@ func (p *Plugin) submitCreateCloudInstance(f *flow.Flow, submission map[string]i
...
+	}
+	jiraURL = strings.TrimSpace(jiraURL)
+	if jiraOrgRegexp.MatchString(jiraURL) {
+		jiraURL = fmt.Sprintf("https://%s.atlassian.net", jiraURL)
+	}
+
+	clientID, _ := submission["client_id"].(string)
+	if clientID == "" {
+		return "", nil, nil, errors.New("no Jira OAuth Client ID in the request")
```
Comment (Kshitij-Katiyar) [inline server/setup_flow.go:613]: Check this everywhere
```suggestion
		return "", nil, nil, errors.New("no Jira OAuth Client ID present in the request")
```

File: server/user_cloud_oauth.go [hunk 22]
```diff
@@ -0,0 +1,116 @@
...
+	code := r.URL.Query().Get("code")
+	if code == "" {
+		return respondErr(w, http.StatusBadRequest, errors.New("Bad request: missing code"))
+	}
+	state := r.URL.Query().Get("state")
+	if state == "" {
+		return respondErr(w, http.StatusBadRequest, errors.New("Bad request: missing state"))
+	}
+
+	if len(strings.Split(state, "_")) != 2 || strings.Split(state, "_")[1] == "" {
```
Comment (Kshitij-Katiyar) [inline server/user_cloud_oauth.go:30]: split the string and put that in a variable and use that everywhere

File: server/user_cloud_oauth.go [hunk 23]
```diff
@@ -0,0 +1,116 @@
...
+		return respondErr(w, http.StatusUnauthorized, errors.New("state not found or might be expired"))
+	}
+	parsed := strings.Split(storedSecret, "_")
+	if len(parsed) < 2 || parsed[0] != stateSecret {
+		return respondErr(w, http.StatusUnauthorized, errors.New("state token mismatch"))
+	}
+
+	mmUser, appErr := p.API.GetUser(mattermostUserID)
+	if appErr != nil {
+		return respondErr(w, http.StatusInternalServerError, errors.WithMessage(appErr, "failed to load user "+mattermostUserID))
```
Comment (Kshitij-Katiyar) [inline server/user_cloud_oauth.go:47]: dont use + for concentation

File: server/user_cloud_oauth.go [hunk 24]
```diff
@@ -0,0 +1,116 @@
...
+	if err != nil {
+		return http.StatusInternalServerError, err
+	}
+
+	client, err := instance.GetClient(connection)
+	if err != nil {
+		return respondErr(w, http.StatusInternalServerError, err)
+	}
+
+	juser, err := client.GetSelf()
```
Comment (Kshitij-Katiyar) [inline server/user_cloud_oauth.go:65]:
```suggestion
	jiraUser, err := client.GetSelf()
```

File: server/user_cloud_oauth.go [hunk 25]
```diff
@@ -0,0 +1,116 @@
...
+	if err != nil {
+		return respondErr(w, http.StatusInternalServerError, err)
+	}
+	connection.User = *juser
+
+	// Set default settings the first time a user connects
+	connection.Settings = &ConnectionSettings{Notifications: true}
+	connection.MattermostUserID = types.ID(mattermostUserID)
+
+	err = p.connectUser(instance, types.ID(mattermostUserID), connection)
```
Comment (Kshitij-Katiyar) [inline server/user_cloud_oauth.go:75]: same

File: server/user_cloud_oauth.go [hunk 26]
```diff
@@ -0,0 +1,116 @@
...
+		JiraDisplayName       string
+		RevokeURL             string
+	}{
+		JiraDisplayName:       juser.DisplayName + " (" + juser.Name + ")",
+		MattermostDisplayName: mmUser.GetDisplayName(model.ShowNicknameFullName),
+		RevokeURL:             path.Join(p.GetPluginURLPath(), instancePath(routeUserDisconnect, instance.GetID())),
+	})
+}
+
+func (p *Plugin) GenerateInitialOAuthToken(mattermostUserID string, instanceID types.ID, code string) (*Connection, error) {
```
Comment (Kshitij-Katiyar) [inline server/user_cloud_oauth.go:91]: same

File: assets/templates/oauth2/complete.html [hunk 27]
```diff
@@ -0,0 +1,81 @@
...
+<html lang="en">
+	<head>
+		<script>
+			window.open('','_parent','');
+			window.close();
+		</script>
+		<style>
+			body {
+				color: rgb(23, 43, 77);
+				letter-spacing: -0.01em;
```
Comment (ayusht2810) [inline assets/templates/oauth2/complete.html:11]: avoid negative css if possible

File: assets/templates/oauth2/complete.html [hunk 28]
```diff
@@ -0,0 +1,81 @@
...
+			}
+
+			.btn {
+				-webkit-transition: all 0.15s ease;
+				-webkit-transition-delay: 0s;
+				-moz-transition: all 0.15s ease;
+				-o-transition: all 0.15s ease;
+				transition: all 0.15s ease false;
+				padding-right: 1em;
+				padding-left: 1em;
```
Comment (ayusht2810) [inline assets/templates/oauth2/complete.html:25]: we can use shorthand property here

File: assets/templates/oauth2/complete.html [hunk 29]
```diff
@@ -0,0 +1,81 @@
...
+			}
+
+			.btn-primary:hover,
+			.btn-primary:active {
+				background: rgb(0, 101, 255);
+			}
+
+			.btn-link {
+				color: #505f79;
+				background: #f4f5f7;
```
Comment (ayusht2810) [inline assets/templates/oauth2/complete.html:45]: I think we should use `rgb` everywhere to maintain consistency. fix everywhere, if updating

File: assets/templates/oauth2/complete.html [hunk 30]
```diff
@@ -0,0 +1,81 @@
...
+				padding: 10px;
+			}
+
+			.btn-link:hover,
+			.btn-link:active {
+				background: #ebecf0;
+			}
+
+			.accounts-container {
+				padding: 1.6em 0 0.8em;
```
Comment (ayusht2810) [inline assets/templates/oauth2/complete.html:55]: same for these units

File: server/instance_cloud_oauth.go [hunk 31]
```diff
@@ -0,0 +1,209 @@
...
+	"golang.org/x/oauth2"
+
+	"github.com/mattermost/mattermost-plugin-jira/server/utils"
+	"github.com/mattermost/mattermost-plugin-jira/server/utils/types"
+)
+
+type cloudOAuthInstance struct {
+	*InstanceCommon
+
+	// The SiteURL may change as we go, so we store the PluginKey when as it was installed
```
Comment (ayusht2810) [inline server/instance_cloud_oauth.go:23]:
```suggestion
	// The SiteURL may change as we go, so we store the PluginKey when it was installed
```

File: server/instance_cloud_oauth.go [hunk 32]
```diff
@@ -0,0 +1,209 @@
...
+		return "", nil, err
+	}
+
+	return jiraURL, instance, err
+}
+
+func (ci *cloudOAuthInstance) GetClient(connection *Connection) (Client, error) {
+	client, _, err := ci.getClientForConnection(connection)
+	if err != nil {
+		return nil, errors.WithMessage(err, "failed to get Jira client for user "+connection.DisplayName)
```
Comment (ayusht2810) [inline server/instance_cloud_oauth.go:79]:
```suggestion
		return nil, errors.WithMessage(err, "failed to get Jira client for the user "+connection.DisplayName)
```

File: server/instance_cloud_oauth.go [hunk 33]
```diff
@@ -0,0 +1,209 @@
...
+	return newCloudClient(client), nil
+}
+
+func (ci *cloudOAuthInstance) getClientForConnection(connection *Connection) (*jira.Client, *http.Client, error) {
+	oauth2Conf := ci.GetOAuthConfig()
+	ctx := context.Background()
+	tokenSource := oauth2Conf.TokenSource(ctx, connection.OAuth2Token)
+	client := oauth2.NewClient(ctx, tokenSource)
+
+	// Get a new token if Access Token has expired
```
Comment (ayusht2810) [inline server/instance_cloud_oauth.go:90]:
```suggestion
	// Get a new token, if Access Token has expired
```

File: server/instance_cloud_oauth.go [hunk 34]
```diff
@@ -0,0 +1,209 @@
...
+	currentToken := connection.OAuth2Token
+	updatedToken, err := tokenSource.Token()
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "error getting token from token source")
```
Comment (ayusht2810) [inline server/instance_cloud_oauth.go:94]:
```suggestion
		return nil, nil, errors.Wrap(err, "error in getting token from token source")
```

File: server/instance_cloud_oauth.go [hunk 35]
```diff
@@ -0,0 +1,209 @@
...
+	}
+
+	if updatedToken.RefreshToken != currentToken.RefreshToken {
+		connection.OAuth2Token = updatedToken
+
+		// Store this new access token & refresh token to get a new access token in future when it has expired
```
Comment (ayusht2810) [inline server/instance_cloud_oauth.go:100]:
```suggestion
		// Store this new access token & refresh the token to get a new access token in the future when it has expired
```

File: server/instance_cloud_oauth.go [hunk 36]
```diff
@@ -0,0 +1,209 @@
...
+
+func (ci *cloudOAuthInstance) GetUserConnectURL(mattermostUserID string) (string, *http.Cookie, error) {
+	oauthConf := ci.GetOAuthConfig()
+	state := fmt.Sprintf("%s_%s", model.NewId()[0:15], mattermostUserID)
+	url := oauthConf.AuthCodeURL(
+		state,
+		oauth2.SetAuthURLParam("audience", "api.atlassian.com"),
+		oauth2.SetAuthURLParam("state", state),
+		oauth2.SetAuthURLParam("response_type", "code"),
+		oauth2.SetAuthURLParam("prompt", "consent"),
```
Comment (ayusht2810) [inline server/instance_cloud_oauth.go:132]: I think there were constants for this. check if we can use them here

File: server/instance_cloud_oauth.go [hunk 37]
```diff
@@ -0,0 +1,209 @@
...
+	return ci.JiraBaseURL
+}
+
+func (ci *cloudOAuthInstance) GetManageAppsURL() string {
+	// TODO
+	return fmt.Sprintf("%s/plugins/servlet/applinks/listApplicationLinks", ci.GetURL())
+}
+
+func (ci *cloudOAuthInstance) GetManageWebhooksURL() string {
+	// TODO
```
Comment (ayusht2810) [inline server/instance_cloud_oauth.go:167]: same

File: server/instance_cloud_oauth.go [hunk 38]
```diff
@@ -0,0 +1,209 @@
...
+	return fmt.Sprintf("%s/plugins/servlet/webhooks", ci.GetURL())
+}
+
+func (ci *cloudOAuthInstance) GetMattermostKey() string {
+	return ci.MattermostKey
+}
+
+func (ci *cloudOAuthInstance) getJiraCloudResourceID(client http.Client) (string, error) {
+	request, err := http.NewRequest(
+		"GET",
```
Comment (ayusht2810) [inline server/instance_cloud_oauth.go:177]: use constant from the package

File: server/instances.go [hunk 39]
```diff
@@ -115,6 +115,17 @@ func (instances Instances) isAliasUnique(instanceID types.ID, alias string) (boo
 	return true, ""
 }
 
+// checkIfExists returns true if the specified instance id already exists
```
Comment (ayusht2810) [inline server/instances.go:118]:
```suggestion
// checkIfExists returns true if the specified instance ID already exists
```

File: server/instances_test.go [hunk 40]
```diff
@@ -110,7 +110,7 @@ func TestInstallInstance(t *testing.T) {
 
 			testInstance0 := &testInstance{
 				InstanceCommon: InstanceCommon{
-					InstanceID: mockInstance1URL,
+					InstanceID: mockInstance3URL,
```
Comment (ayusht2810) [inline server/instances_test.go:113]: why this change?

File: server/setup_flow.go [hunk 41]
```diff
@@ -87,6 +85,25 @@ func (p *Plugin) NewSetupFlow() *flow.Flow {
 		InitHTTP(p.router)
 }
 
+func (p *Plugin) NewOAuth2Flow() *flow.Flow {
+	pluginURL := *p.client.Configuration.GetConfig().ServiceSettings.SiteURL + "/" + "plugins" + "/" + Manifest.Id
```
Comment (ayusht2810) [inline server/setup_flow.go:89]:
```suggestion
	pluginURL := *p.client.Configuration.GetConfig().ServiceSettings.SiteURL + "/plugins/" + Manifest.Id
```

File: server/setup_flow.go [hunk 42]
```diff
@@ -271,44 +289,56 @@ func (p *Plugin) stepServerConfigureAppLink2() flow.Step {
...
+			"2. Name your app according to its purpose, for example:  `Mattermost Jira Plugin - <your company name>`.\n"+
+			"3. Accept the **Terms** and click **Create**.\n"+
+			"4. Select **Permissions** in the left menu. Next to the JIRA API, select **Add**.\n"+
+			"5. Then select **Configure** and ensure following scopes are selected:\n"+
+			"       %s\n"+
+			"6. Select **Authorization** in the left menu.\n"+
+			"7. Next to OAuth 2.0 (3LO), select **Add** and set the Callback URL as follows and click **Save Changes**:\n"+
+			"       {{.OAuthCompleteURL}}\n"+
+			"8. Select **Settings** in the left menu.\n"+
+			"9. Copy the **Client ID** and **Secret** and keep it handy.\n"+
```
Comment (ayusht2810) [inline server/setup_flow.go:306]: check if this is `Client Secret` or just `Secret`

File: server/user.go [hunk 43]
```diff
@@ -28,10 +29,12 @@ type User struct {
 type Connection struct {
 	jira.User
 	PluginVersion      string
-	Oauth1AccessToken  string `json:",omitempty"`
-	Oauth1AccessSecret string `json:",omitempty"`
+	Oauth1AccessToken  string        `json:",omitempty"`
+	Oauth1AccessSecret string        `json:",omitempty"`
+	OAuth2Token        *oauth2.Token `json:",omitempty"`
```
Comment (ayusht2810) [inline server/user.go:34]: just a question is `,` required here?

File: server/user_cloud_oauth.go [hunk 44]
```diff
@@ -0,0 +1,116 @@
...
+	}
+
+	mmUser, appErr := p.API.GetUser(mattermostUserID)
+	if appErr != nil {
+		return respondErr(w, http.StatusInternalServerError, errors.WithMessage(appErr, "failed to load user "+mattermostUserID))
+	}
+
+	instance, err := p.instanceStore.LoadInstance(instanceID)
+	if err != nil {
+		return http.StatusInternalServerError, err
```
Comment (ayusht2810) [inline server/user_cloud_oauth.go:52]: why are we not using `respondErr` here? update everywhere, if changing

File: server/user_cloud_oauth.go [hunk 45]
```diff
@@ -0,0 +1,116 @@
...
+		return respondErr(w, http.StatusInternalServerError, err)
+	}
+
+	juser, err := client.GetSelf()
+	if err != nil {
+		return respondErr(w, http.StatusInternalServerError, err)
+	}
+	connection.User = *juser
+
+	// Set default settings the first time a user connects
```
Comment (ayusht2810) [inline server/user_cloud_oauth.go:71]:
```suggestion
	// Set default settings when the user connects for the first time
```

File: server/instance_cloud_oauth.go (hunk 16 above)
Comment (raghavaggarwal2308) [reply server/instance_cloud_oauth.go:203]: I think what we haveis better

File: server/setup_flow.go (hunk 19 above)
Comment (raghavaggarwal2308) [reply server/setup_flow.go:602]: its not an error

File: server/instance_cloud_oauth.go (hunk 10 above)
Comment (raghavaggarwal2308) [reply server/instance_cloud_oauth.go:176]: There is no such method

File: server/instance_cloud_oauth.go (hunk 35 above)
Comment (raghavaggarwal2308) [reply server/instance_cloud_oauth.go:99]: I think what we have is fine

File: server/instances_test.go (hunk 40 above)
Comment (raghavaggarwal2308) [reply server/instances_test.go:113]: Testcases are failing without this change. I think there are some changes in the function

File: server/setup_flow.go (hunk 42 above)
Comment (raghavaggarwal2308) [reply server/setup_flow.go:312]: Its `Secret`

File: server/user.go (hunk 43 above)
Comment (raghavaggarwal2308) [reply server/user.go:34]: I think yes. THe perm before `,` is the json name of the field

File: assets/templates/oauth2/complete.html (hunk 30 above)
Comment (raghavaggarwal2308) [reply assets/templates/oauth2/complete.html:55]: discussed

File: server/setup_flow.go [hunk 46]
```diff
@@ -554,15 +590,51 @@ func (p *Plugin) submitCreateCloudInstance(f *flow.Flow, submission map[string]i
...
+	jiraURL, _ := submission[NameURL].(string)
+	if jiraURL == "" {
+		return "", nil, nil, errors.New("no Jira cloud URL is present in the request")
+	}
+	jiraURL = strings.TrimSpace(jiraURL)
+	if jiraOrgRegexp.MatchString(jiraURL) {
+		jiraURL = fmt.Sprintf("https://%s.atlassian.net", jiraURL)
+	}
+
+	clientID, _ := submission[NameClientID].(string)
```
Comment (Kshitij-Katiyar) [inline server/setup_flow.go:617]: same

File: server/setup_flow.go [hunk 47]
```diff
@@ -554,15 +590,51 @@ func (p *Plugin) submitCreateCloudInstance(f *flow.Flow, submission map[string]i
...
 
-	return stepCloudAddedInstance, flow.State{
-		keyEdition:             string(CloudInstanceType),
-		keyJiraURL:             jiraURL,
-		keyAtlassianConnectURL: p.GetPluginURL() + instancePath(routeACJSON, types.ID(jiraURL)),
+	return stepInstalledJiraApp, flow.State{
+		keyEdition:          string(CloudOAuthInstanceType),
+		keyJiraURL:          jiraURL,
+		keyInstance:         instance,
+		keyOAuthCompleteURL: p.GetPluginURL() + instancePath(routeOAuth2Complete, types.ID(jiraURL)),
```
Comment (Kshitij-Katiyar) [inline server/setup_flow.go:636]: use fmt.sprintf

File: server/user_cloud_oauth.go [hunk 48]
```diff
@@ -0,0 +1,117 @@
...
+	}
+
+	mmUser, appErr := p.API.GetUser(mattermostUserID)
+	if appErr != nil {
+		return respondErr(w, http.StatusInternalServerError, errors.WithMessage(appErr, fmt.Sprintf("failed to load user %s", mattermostUserID)))
+	}
+
+	instance, err := p.instanceStore.LoadInstance(instanceID)
+	if err != nil {
+		return respondErr(w, http.StatusInternalServerError, err)
```
Comment (Kshitij-Katiyar) [inline server/user_cloud_oauth.go:54]: Wrap all the errors here

File: assets/templates/oauth2/complete.html [hunk 49]
```diff
@@ -1,10 +1,6 @@
 <!DOCTYPE html>
 <html lang="en">
 	<head>
-		<script>
-			window.open('','_parent','');
-			window.close();
-		</script>
```
[the remark this answers is not in the document]
Comment (raghavaggarwal2308) [reply assets/templates/oauth2/complete.html]: discussed

[the remark this answers is not in the document]
Comment (raghavaggarwal2308) [reply assets/templates/oauth2/complete.html]: discussed

**Review by Kshitij-Katiyar — APPROVED** [review]

**Review by ayusht2810 — APPROVED** [review]

**Review by Kshitij-Katiyar — APPROVED** [review]
